torch/csrc/jit/tensorexpr/schedule.cpp

```diff
@@ -873,6 +873,109 @@ LoopAxis* LoopAxisTransform::NewAxis(
   return axis;
 }
 
+// XXX
+LoopNest::LoopNest(const std::vector<Tensor*> tensors_to_compute) {
+  std::vector<Tensor*> output_tensors(tensors_to_compute);
+
+  std::vector<Stmt*> loops;
+  for (Tensor *t : tensors_to_compute) {
+    Stmt* loop = LowerToStmt(t);
+    loops.push_back(loop);
+  }
+  root_stmt_ = new Block(loops);
+}
+
+Stmt* LoopNest::LowerToStmt(Tensor* t) {
+  Function* f = t->function();
+  // TODO: Support multiple-output functions
+  Stmt* body = f->ElementStmt(0);
+
+  stmt_to_tensor_[body] = t;
+  tensor_to_stmt_[t] = body;
+
+  CHECK(f->ndim() >= 1);
+  for (size_t i = 0; i < f->ndim(); i++) {
+    // Going in reverse order: from innermost loop to the outermost
+    size_t dim_index = f->ndim() - i - 1;
+    Range r(0, ExprHandle(f->dim(dim_index)));
+    body = For::make(
+        VarHandle(f->arg(dim_index)), r.start(), r.stop(), body);
+  }
+  return body;
+}
+
+void LoopNest::ComputeInline(Stmt *s) {
+  // TODO: check if `s` is a body of a loop
+  inlined_functions_.insert(stmt_to_tensor_.at(s)->function());
+}
+
+void LoopNest::ApplyInlines() {
+  // TODO: check if `s` is a body of a loop
+  std::vector<Function*> inlined_functions_vec(
+      inlined_functions_.begin(), inlined_functions_.end());
+  root_stmt_ = InjectInlines(root_stmt_, inlined_functions_vec);
+}
+
+void LoopNest::SplitWithTail(Stmt *s, int factor, Stmt** inner, Stmt **outer, Stmt **tail) {
```

Comment:
Why do we give inner/outer back to user?

Reply:
It's optional, I just wanted to keep the API as close as possible to the existing one. In practice I'd expect this to be useful for chaining transformations, e.g.:
```
Stmt *j_outer, *j_inner, *j_tail;
l.SplitWithTail(loops[2], 17, &j_outer, &j_inner, &j_tail);
l.Vectorize(j_inner); // This is not implemented here now
```
One could get to the same statement via `getLoopStmtsFor` API, but having it ready right after `SplitWithTail` might be convenient.

Reply:
even fluent interface? :)